### PR TITLE
Improvements to Array#sample

### PR DESCRIPTION
**Pull request by LillianZ**
On average this PR doubles iterations per second of `Array#sample`.

Minor correctness fix in sample on error message and count = 2 .

Mainly fixed and improved the previous non-swap implementation of `sample_many` (named quad in this PR), added a new implementation using hash, improved logic for picking an implementation by experimentally determining which is best when.

Benchmark array_sample_m_n is
```
arr = [*0...m]
c.times {arr.sample n} 
```
where c is probably 1_000 (I forget).

Ips from benchmark-ips from before and after this PR (on the master of 2.5 months ago); 

|                       | Old ips |          | New ips |          |
|-----------------------|---------|----------|---------|----------|
| array_sample_100_3    | 5.503k  | (± 3.9%) | 22.432k | (±12.5%) |
| array_sample_100_8    | 8.252k  | (± 2.0%) | 31.619k | (± 8.5%) |
| array_sample_100_10   | 2.986k  | (± 1.8%) | 17.716k | (± 3.1%) |
| array_sample_100_12   | 9.212k  | (± 3.2%) | 9.704k  | (± 2.5%) |
| array_sample_100_15   | 2.887k  | (± 4.2%) | 17.734k | (± 1.8%) |
| array_sample_100_30   | 9.292k  | (± 2.1%) | 9.766k  | (± 1.1%) |
| array_sample_100_60   | 4.996k  | (± 1.9%) | 5.152k  | (± 1.2%) |
| array_sample_1k_10    | 17.359k | (± 3.2%) | 25.839k | (±16.8%) |
| array_sample_1k_11    | 15.278k | (± 2.3%) | 23.815k | (±12.0%) |
| array_sample_1k_30    | 2.966k  | (± 1.6%) | 6.880k  | (±10.2%) |
| array_sample_1k_50    | 1.382k  | (± 1.4%) | 4.670k  | (± 6.2%) |
| array_sample_1k_70    | 988.056 | (± 1.1%) | 3.721k  | (± 5.1%) |
| array_sample_1k_100   | 825.393 | (± 2.7%) | 2.964k  | (± 3.5%) |
| array_sample_1k_600   | 583.461 | (± 2.2%) | 596.062 | (± 1.3%) |
| array_sample_10k_10   | 17.942k | (± 3.1%) | 18.592k | (±13.9%) |
| array_sample_10k_11   | 16.122k | (± 1.8%) | 16.914k | (±11.7%) |
| array_sample_10k_30   | 4.089k  | (± 6.4%) | 5.328k  | (± 9.4%) |
| array_sample_10k_80   | 887.231 | (± 4.1%) | 1.691k  | (± 4.3%) |
| array_sample_10k_200  | 253.053 | (± 4.7%) | 692.278 | (± 5.6%) |
| array_sample_10k_500  | 171.113 | (± 4.1%) | 388.098 | (± 2.3%) |
| array_sample_10k_1k   | 126.663 | (± 3.9%) | 219.515 | (± 2.7%) |
| array_sample_10k_6k   | 41.194  | (± 0.0%) | 41.807  | (± 2.4%) |
| array_sample_100k_10  | 20.509k | (± 1.9%) | 20.840k | (± 9.7%) |
| array_sample_100k_11  | 18.211k | (± 1.9%) | 19.543k | (± 7.2%) |
| array_sample_100k_30  | 5.162k  | (± 1.8%) | 6.098k  | (± 2.8%) |
| array_sample_100k_60  | 1.839k  | (± 2.1%) | 2.394k  | (± 2.3%) |
| array_sample_100k_70  | 1.434k  | (± 3.1%) | 1.909k  | (± 1.7%) |
| array_sample_100k_80  | 1.154k  | (± 2.2%) | 1.118k  | (± 2.8%) |
| array_sample_100k_200 | 243.148 | (± 2.5%) | 440.921 | (± 3.4%) |
| array_sample_100k_500 | 60.461  | (± 3.3%) | 169.728 | (± 2.9%) |
| array_sample_100k_1k  | 40.006  | (± 2.5%) | 83.773  | (± 3.6%) |
| array_sample_100k_4k  | 28.380  | (± 3.5%) | 50.109  | (± 2.0%) |
| array_sample_100k_5k  | 25.733  | (± 3.9%) | 42.937  | (± 2.3%) |
| array_sample_100k_6k  | 23.666  | (± 4.2%) | 37.777  | (± 2.6%) |
| array_sample_100k_10k | 17.922  | (± 5.6%) | 25.308  | (± 4.0%) |
| array_sample_100k_50k | 5.918   | (± 0.0%) | 5.877   | (± 0.0%) |
| array_sample_1M_10    | 21.742k | (± 8.8%) | 23.268k | (± 9.1%) |
| array_sample_1M_11    | 18.796k | (± 9.1%) | 20.073k | (± 8.2%) |
| array_sample_1M_30    | 4.781k  | (± 9.2%) | 5.987k  | (±11.4%) |
| array_sample_1M_80    | 1.018k  | (± 8.1%) | 1.068k  | (± 8.8%) |
| array_sample_1M_100   | 755.304 | (± 8.9%) | 808.367 | (± 8.5%) |
| array_sample_1M_130   | 484.768 | (±10.5%) | 616.955 | (±10.7%) |
| array_sample_1M_180   | 277.801 | (± 9.0%) | 424.511 | (± 8.2%) |
| array_sample_1M_200   | 209.531 | (± 7.2%) | 374.245 | (± 9.9%) |
| array_sample_1M_500   | 41.065  | (± 7.3%) | 139.475 | (± 2.2%) |
| array_sample_1M_1k    | 12.615  | (± 7.9%) | 69.492  | (± 1.4%) |
| array_sample_1M_6k    | 4.700   | (± 0.0%) | 17.081  | (± 0.0%) |
| array_sample_1M_10k   | 13.254  | (± 7.5%) | 127.741 | (± 7.0%) |
| array_sample_1M_50k   | 2.871   | (± 0.0%) | 4.218   | (± 0.0%) |
| array_sample_1M_100k  | 1.883   | (± 0.0%) | 2.676   | (± 0.0%) |

### CODE REVIEW

**Comment by chrisseaton** [comment]
https://github.com/Shopify/truffleruby/issues/1

**Comment by eregon** [comment]
The `swap` implementation looks really simple and efficient, are there cases where it significantly underperforms the other implementations?

File: src/main/ruby/truffleruby/core/array.rb [hunk 1]
```diff
@@ -1007,67 +1007,93 @@ def sample(count=undefined, options=undefined)
...
       i = 1
 
       result[0] = rng.rand(size)
-      while i < count
-        k = rng.rand(size)
 
-        spin = false
-        spin_count = 0
+      if quad #quadratic time due to linear time collision check but low overhead
+        while i < count
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/array.rb:1040]: Could you split each implementation in a separate method for readability?

File: src/main/ruby/truffleruby/core/array.rb [hunk 2]
```diff
@@ -1007,67 +1007,93 @@ def sample(count=undefined, options=undefined)
...
-          while j < i
-            if k == result[j]
-              spin = true
-              break
-            end
+          spin = false
 
-            j += 1
-          end
+          while true # rubocop:disable Lint/LiteralAsCondition
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/array.rb:1045]: (not for this PR, I'll do it)
I'm thinking we should just disable this rule globally, it seems annoying and very rarely useful.
There seem to be no way to configure it only for `if/unless` and not `while/until`.

**Review by eregon — COMMENTED** [review]
Looks great!

**Comment by LillianZ** [comment]
> The `swap` implementation looks really simple and efficient, are there cases where it significantly underperforms the other implementations? Probably when sampling few elements out of a huge array?

Exactly. I'm not sure why (if you know please share), but the swap function is linearly dependant on the size of the array. So for the most extreme example   array_sample_1M_10, 

| Swap | 0.123 ips   |
|------|-------------|
| Hash | 140.212 ips |
| Quad | 280.007 ips |

File: src/main/ruby/truffleruby/core/array.rb (hunk 1 above)
Comment (LillianZ) [reply src/main/ruby/truffleruby/core/array.rb:1040]: done!

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
This PR appears to fail specs (which we should add to GitHub Actions to catch it more easily):
```
1)
Array#sample samples evenly FAILED
Expected 800..1200 to include 0
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:13:in `block (4 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:12:in `each'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:12:in `block (3 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:7:in `block (2 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:4:in `<top (required)>'

2)
Array#sample returns an Array of elements when passed a count FAILED
Expected 3 (Integer) to be an instance of Array
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:31:in `block (2 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:4:in `<top (required)>'

3)
Array#sample returns elements from the Array ERROR
NoMethodError: undefined method `all?' for 3:Integer
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:36:in `block (2 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:4:in `<top (required)>'

4)
Array#sample returns at most the number of elements in the Array FAILED
Expected 8 == 4
to be truthy but was false
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:42:in `block (2 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:4:in `<top (required)>'

5)
Array#sample does not return the same value if the Array has unique values ERROR
NoMethodError: undefined method `sort' for 4:Integer
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:48:in `block (2 levels) in <top (required)>'
/b/b/e/main/spec/ruby/core/array/sample_spec.rb:4:in `<top (required)>'
```

I made a couple commits on top, could you include them in this PR and commit fixes on top to avoid conflicts?
```
git remote add eregon https://github.com/eregon/truffleruby.git
git fetch eregon
git merge --ff-only eregon/ArraySample
```

**Comment by LillianZ** [comment]
I lost a line during refactoring; it's fixed now. Apologies and thanks!

**Pull request merged**